stdlib/Std/src/Base.luna

```diff
@@ -212,6 +212,11 @@ native class Text:
 
     def addPathSegment segment: self + pathSeparator + segment
 
+    # Append a path segment to the given `Text`. The new segment will use
+    # a path separator specific for current platform, making it the safest
+    # way of consturcting paths.
```

Comment:
typo:
consturcting -> constructing

Reply:
Also:
1) When reading this documentation it is not clear to me what "path segment" is. 
2) we should check for trailing / leading separator in path parts when concatenating. Otherwise, we might end up with ugly paths like "foo/\bar"

(Multiple / mixed separators are not really dangerous, just ugly. But if we recommend to user to use this function because of that, then we should keep them tidy.)

Reply:
The issue with something like that is that it precludes paths coming in from other programs or via FFI. I agree that we want to have the 'one true path' format internally, but it's very important we handle this carefully. We likely can't do it right until we have the new type-system and its related conversion mechanisms.

Reply:
@wdanilo While I agree with the reasoning, I disagree with the conclusions. As a Windows user there is nothing I'm more afraid of, than libraries trying to be "platform independent" by trying to force Linux conventions on Windows. I want programs to be portable, i.e. do sensible things on all platforms rather then "behaving exactly the same" at the cost of user-experience.

1. For me as a user, the software is broken, if it cannot accept the paths that my OS presents to me, or if I'm unable to use the paths in other programs in my OS. I need to be able to just copy path from / to Windows explorer into Luna code.

Btw, I remember that people in our team (long time ago, Cabal-hell era) when attempting to fix build on Windows were trying running commands while changing the separators to native. That's just to reinforce how "unreliable" non-native separators might feel just by the look, even if they worked. For Windows-only people it will be only worse.

2. When compared to issues above, actually converting separators at the bottom of IO functions is not that relevant (though should be done nevertheless). System IO API (and most of filesystem libraries) supports both kinds of separators. It is user input/output that needs to present them in a native way. Not all Windows programs correctly support non-native separators input. Interoperability may be seriously hampered if we try to force POSIX conventions onto Window-based users.

***

> we should assume that all of our libraries use just one convention for paths

I would put that in a different way — all our libraries shouldn't really know anything about path conventions. They should just use our `FilePath` type and its methods and that should be the only convention library code needs to use.

No one else except for implementation of FilePath should ever be required to write `/` or `\`. 
`FilePath` should provide methods both for accessing path as native-like string and as a string normalized to `/` separator (i.e. the platform-independent printing).

(and perhaps FilePath shouldn't be string but a sequence of segments and as such never contain any separator characters)

> Users should be able to write paths by hand like "foo/bar/baz" and should be sure they work OK.

Absolutely. But I equally strongly feel that on Windows the Windows-style paths must work as well. Perhaps on Linux as well.

Yes, I see potential issues caused by accepting both path separators. For example, if user types `foo\bar` then it might mean different things depending on platform. `\` is an allowed character in path on Linux, so the whole thing would be a single filename. But `$` is allowed as well. This is just the 0.1% (well, in fact way less — how many files with `\` are on your filesystem?) that won't implicitly work out of the box.
Otherwise… I think we would be fine.

Reply:
OK, this PR is not about revamping how we handle paths. So, how about, for the purpose of this PR:
1. We resolve the issues around wording in the documentation.
2. We leave the behavior as is.
3. We let the design for proper filepaths handling crystalize to its final form and then do that as an effort on its own?

My rationale would be: this functionality here has nothing to do with the way filepaths behave. We're going on a tangent here and the fact of the matter is: this PR changes nothing in the way paths behave. @wdanilo @iamrecursion @mwu-tow you with me on this one?

Reply:
Definitely. This PR needs to go through, and we can take our time to redesign paths to be sensible and work with the new type discipline.

Reply:
@kustosz Yes, absolutely. Sorry for hijacking the PR for design discussion. 

I think that my issue with having leading/trailing separators in segments merged by `/` still stands, no matter what we decide regarding the separator selection itself. I think that `"foo/" / "bar"` should be on Linux `"foo/bar"`, not "foo//bar"`.  Not a blocker, but an issue nevertheless.

Reply:
https://github.com/luna/luna/issues/399